File: src/main/java/org/zbus/rpc/Rpc.java
package org.zbus.rpc;


import java.io.IOException;

import org.zbus.log.Logger;
import org.zbus.mq.Broker;
import org.zbus.net.http.Message;
import org.zbus.rpc.service.Caller;

public class Rpc extends Caller{  
	private static final Logger log = Logger.getLogger(Rpc.class);
	private static final Codec codec = new JsonCodec();
	public static final String DEFAULT_ENCODING = "UTF-8";  
	
	private String module = ""; 
	private String encoding = DEFAULT_ENCODING;
	private int timeout = 10000;  
	private boolean verbose = false;

	public Rpc(Broker broker, String mq) {
		super(broker, mq); 
	}
	
	public Rpc(RpcConfig config){
		super(config);
		this.module = config.getModule();
		this.timeout = config.getTimeout(); 
		this.encoding = config.getEncoding();
		this.verbose = config.isVerbose();
	}
	
	@SuppressWarnings("unchecked")
	public <T> T invokeSync(Class<T> clazz, String method, Object... args){
		Object netObj = invokeSync(method, args);
		try {
			return (T) codec.normalize(netObj, clazz);
		} catch (ClassNotFoundException e) { 
			throw new RpcException(e.getMessage(), e.getCause());
		}
	}
	
	@SuppressWarnings("unchecked")
	public <T> T invokeSyncWithType(Class<T> clazz, String method, Class<?>[] types, Object... args){
		Object netObj = invokeSyncWithType(method, types, args);
		try {
			return (T) codec.normalize(netObj, clazz);
		} catch (ClassNotFoundException e) { 
			throw new RpcException(e.getMessage(), e.getCause());
		}
	}
	
	public Object invokeSync(String method, Object... args) {
		return invokeSyncWithType(method, null, args);
	} 
	
	public Object invokeSyncWithType(String method, Class<?>[] types, Object... args) {	
		Request req = new Request();
		req.setModule(this.module);
		req.setMethod(method); 
		req.setParams(args); 
		req.assignParamTypes(types); 
		req.setEncoding(this.encoding);
		 
		Message msgReq= null, msgRes = null;
		try {
			long start = System.currentTimeMillis();
			msgReq = codec.encodeRequest(req); 
			if(isVerbose()){
				log.info("[REQ]: %s", msgReq);
			} 
			
			msgRes = this.invokeSync(msgReq, this.timeout); 
			
			if(isVerbose()){
				long end = System.currentTimeMillis();
				log.info("[REP]: Time cost=%dms\n%s",(end-start), msgRes);
			} 
			
		} catch (IOException e) {
			throw new RpcException(e.getMessage(), e);
		}
		
		if (msgRes == null) { 
			String errorMsg = String.format("MQ(%s)-module(%s)-method(%s) request timeout\n%s", 
					mq,  module, method, msgReq.toString());
			throw new RpcException(errorMsg);
		}
		
		Response resp = codec.decodeResponse(msgRes);
		
		
		if(resp.getStackTrace() != null){
			Throwable error = resp.getError();
			if(error != null){
				if(error instanceof RuntimeException){
					throw (RuntimeException)error;
				}
				throw new RpcException(error.getMessage(), error.getCause()); 
			} else {
				throw new RpcException(resp.getStackTrace());
			}
		} 
		return resp.getResult();
	}

	public String getEncoding() {
		return encoding;
	}

	public void setEncoding(String encoding) {
		this.encoding = encoding;
	}

	public String getModule() {
		return module;
	}

	public void setModule(String module) {
		this.module = module;
	}
	
	public Rpc module(String module) {
		this.module = module;
		return this;
	}

	public int getTimeout() {
		return timeout;
	}

	public void setTimeout(int timeout) {
		this.timeout = timeout;
	}

	public boolean isVerbose() {
		return verbose;
	}

	public void setVerbose(boolean verbose) {
		this.verbose = verbose;
	} 
	
}

File: NET/Zbus/Rpc/Rpc.cs
using System;
using System.Collections.Generic;
using System.Text;

using Zbus.Kit.Json;
using Zbus.Mq;
using Zbus.Broker;
using Zbus.Net.Http;

namespace Zbus.RPC
{
   public class Request
   {
      private string mq; //
      private string module;
      private string method;
      private object[] args;
      private object[] argTypes;
      private string encoding = "UTF-8";


      public string Mq
      {
         get { return mq; }
         set { mq = value; }
      }

      public string Module
      {
         get { return module; }
         set { module = value; }
      }
      public string Method
      {
         get { return method; }
         set { method = value; }
      }

      public object[] Args
      {
         get { return args; }
         set { args = value; }
      }

      public object[] ArgTypes
      {
         get { return argTypes; }
         set { argTypes = value; }
      }

      public string Encoding
      {
         get { return encoding; }
         set { encoding = value; }
      }
   }

   public class Rpc
   {
      private IBroker broker;
      private string mq;
      private string token = "";
      private String module = "";
      private String encoding = RpcConfig.DEFAULT_ENCODING;
      private int timeout = 10000;

      public Rpc(IBroker broker)
      {
         this.broker = broker;
      }

      public Rpc(IBroker broker, String mq)
      {
         this.broker = broker;
         this.mq = mq;
      }

      public Rpc(RpcConfig config)
      {
         this.broker = config.Broker;
         this.mq = config.Mq;
         this.token = config.AccessToken;
         this.module = config.Module;
         this.encoding = config.Encoding;
         this.timeout = config.Timeout;
      }

      private Message Invoke(Message msg, int timeout)
      {
         if (msg.Mq == null)
         {
            if (this.mq == null)
            {
               throw new RpcException("Rpc invocation missing mq");
            }
            else
            {
               msg.Mq = this.mq;
            }
         }
         msg.Cmd = Proto.Produce;
         msg.Ack = false;

         return this.broker.InvokeSync(msg, timeout);
      }

      public object Invoke(Request request)
      {
         IDictionary<string, object> req = new Dictionary<string, object>();
         req["module"] = request.Module;
         req["method"] = request.Method;
         req["params"] = request.Args;
         req["paramTypes"] = request.ArgTypes;
         req["encoding"] = request.Encoding;

         Message msgReq = new Message();
         string json = JSON.Instance.ToJSON(req);
         msgReq.SetJsonBody(json);
         if (request.Mq != null)
         {
            msgReq.Mq = request.Mq;
         }


         Message msgRes = this.Invoke(msgReq, this.timeout);
         string encodingName = msgRes.Encoding;
         Encoding encoding = Encoding.Default;
         if (encodingName != null)
         {
            encoding = Encoding.GetEncoding(encodingName);
         }
         string jsonString = msgRes.GetBody(encoding);
         Dictionary<string, object> jsonRes = (Dictionary<string, object>)JSON.Instance.Parse(jsonString);
         if (jsonRes.ContainsKey("error"))
         {
            if (jsonRes["error"] != null)
               throw new RpcException((string)jsonRes["error"]);
         }

         if (jsonRes.ContainsKey("result"))
         {
            return jsonRes["result"];
         }


         throw new RpcException("return format error");
      }

      public object Invoke(string method, params object[] args)
      {
         Request req = new Request();
         req.Module = this.module;
         req.Method = method;
         req.Args = args;
         req.Encoding = this.encoding;

         return Invoke(req);
      }
   }

}
